fairlearn/metrics/_metric_frame.py

```diff
@@ -0,0 +1,515 @@
+
+    def _compute_dataframe_from_rows(self, func_dict, y_true, y_pred, rows):
+        if len(rows) == 1:
+            row_index = pd.Index(data=rows[0].classes, name=rows[0].name)
+        else:
+            row_index = pd.MultiIndex.from_product([x.classes for x in rows],
+                                                   names=[x.name for x in rows])
+
+        if len(row_index) > _SUBGROUP_COUNT_WARNING_THRESHOLD:
+            msg = _SUBGROUP_COUNT_WARNING.format(len(row_index))
+            logger.warning(msg)
+
+        result = pd.DataFrame(index=row_index, columns=func_dict.keys())
+        for func_name in func_dict:
+            for row_curr in row_index:
+                mask = None
+                if len(rows) > 1:
+                    mask = self._mask_from_tuple(row_curr, rows)
+                else:
+                    # Have to force row_curr to be an unary tuple
+                    mask = self._mask_from_tuple((row_curr,), rows)
+
+                # Only call the metric function if the mask is non-empty
+                if sum(mask) > 0:
+                    curr_metric = func_dict[func_name].evaluate(y_true, y_pred, mask)
+                    result[func_name][row_curr] = curr_metric
+        return result
+
+    @property
+    def overall(self) -> Union[pd.Series, pd.DataFrame]:
+        """Return the underlying metrics evaluated on the whole dataset.
+
+        Returns
+        -------
+        pandas.Series or pandas.DataFrame
+            If no ``control_levels`` were specified, then this is
+            a :class:`pandas.Series` with one row for each underlying metric.
+
+            If ``control_levels`` were specified, then this is a
+            :class:`pandas.DataFrame`, with columns corresponding to the
+            underling metric(s) and the rows indexed by the subgroups of
+            the control features.
+        """
+        return self._overall
+
+    @property
+    def by_group(self) -> pd.DataFrame:
+        """Return the collection of metrics evaluated for each subgroup.
+
+        Returns
+        -------
+        pandas.DataFrame
+            The result of evaluating
+            each underlying metric for each combination of classes in the
+            sensitive and control features. The columns identify
+            the underlying metric, while the rows are indexed by the unique
+            combinations of classes found in the sensitive and control
+            features. If control features are present, they are in the
+            'outer' layer of the row indexing.
+
+            If a particular combination of classes was not present in the dataset
+            (likely to occur as more sensitive and control features
+            are specified), then the corresponding entry in the DataFrame will
+            be NaN.
+        """
+        return self._by_group
+
+    @property
+    def control_levels(self) -> List[str]:
```

Comment:
Right now we are returning `[]` when no control features are provided. I think that we should return `None` in that case, so we can write things like
`self.by_group.max(level=self.control_levels)`.

Reply:
That's an interesting point actually. The following two work:
```python
mf.by_group.max(level=None)
mf.by_group.unstack(level=[])
```
The following two fail:
```python
mf.by_group.max(level=[])
mf.by_group.unstack(level=None)
```
So, the question is more 'which set of methods do we prioritise?'

Reply:
I know. I think that it's better to enable `max`, `min`, `mean`, ... than `unstack`.

Reply:
Resolving to `None`.